default/AI/ResearchAI.py

```diff
@@ -20,12 +18,331 @@ def get_research_index():
+    if tech_name in ["SHP_INTSTEL_LOG"]:
+        return DEFAULT # interstellar logistics increases ship speed
+
+    # Robotic hulls
+    if tech_name in ["SHP_MIL_ROBO_CONT", "SHP_MIDCOMB_LOG"]:
+        return robotic
+    if tech_name in ["SHP_SPACE_FLUX_DRIVE", "SHP_TRANSSPACE_DRIVE"]:
+        return min(robotic, flux)
+    if tech_name in ["SHP_CONTGRAV_MAINT", "SHP_MASSPROP_SPEC"]:
+        return min(robotic, contgrav)
+    if tech_name in ["SHP_NANOROBO_MAINT"]:
+        return min(robotic, nanorobo)
+    if tech_name in ["SHP_XENTRONIUM_HULL"]:
+        return DEFAULT # this is not a robotic hull!
+
+    # Asteroid hulls
+    if tech_name in ["SHP_ASTEROID_HULLS", "SHP_ASTEROID_REFORM", "SHP_MONOMOLEC_LATTICE", "SHP_SCAT_AST_HULL"]:
+        return asteroid
+    if tech_name in ["SHP_HEAVY_AST_HULL"]:
+        return min(asteroid, astheavy)
+    if tech_name in ["SHP_CAMO_AST_HULL"]:
+        return min(asteroid, astcamo)
+    if tech_name in ["SHP_MINIAST_SWARM"]:
+        return min(asteroid, astswarm)
+
+    # Organic hulls
+    if tech_name in ["SHP_ORG_HULL", "SHP_SENT_HULL"]:
+        return org
+    if tech_name in ["SHP_MULTICELL_CAST", "SHP_ENDOCRINE_SYSTEMS", "SHP_CONT_BIOADAPT"]:
+        return min(org, orgraven)
+    if tech_name in ["SHP_MONOCELL_EXP", "SHP_CONT_SYMB", "SHP_BIOADAPTIVE_SPEC"]:
+        return min(org, orgneural)
+    if tech_name in ["SHP_ENDOSYMB_HULL"]:
+        return min(org, orgendosym)
+
+    # energy hulls
+    if tech_name in ["SHP_FRC_ENRG_COMP"]:
+        return energy
+    if tech_name in ["SHP_QUANT_ENRG_MAG"]:
+        return min(energy, energymag, IMMEDIATE if has_blue_star or has_black_hole else USELESS)
+    if tech_name in ["SHP_ENRG_BOUND_MAN"]:
+        return min(energy, energyfrac, IMMEDIATE if has_blue_star or has_black_hole else USELESS)
+    if tech_name in ["SHP_SOLAR_CONT"]:
+        return min(energy, IMMEDIATE if has_red_star or has_black_hole else USELESS) # red star can be turned into has_black_hole by Artificial black hole
+
+    # damage control
+    if tech_name in ["SHP_REINFORCED_HULL"]:
+        return armor
+    if tech_name in ["SHP_BASIC_DAM_CONT", "SHP_FLEET_REPAIR", "SHP_ADV_DAM_CONT"]:
+        return repair
+
+    # ship parts
+    if tech_name in ["SHP_IMPROVED_ENGINE_COUPLINGS", "SHP_N_DIMENSIONAL_ENGINE_MATRIX", "SHP_SINGULARITY_ENGINE_CORE"]:
+        return engine
+    if tech_name in ["SHP_DEUTERIUM_TANK", "SHP_ANTIMATTER_TANK", "SHP_ZERO_POINT"]:
+        return fuel
+    if tech_name in ["SHP_NOVA_BOMB", "SHP_DEATH_SPORE", "SHP_BIOTERM"]:
+        return USELESS # AI does not use Nova bomb nor bioterror
+
+    # ship shields
+    if tech_name in ["SHP_DEFLECTOR_SHIELD", "SHP_PLASMA_SHIELD", "SHP_BLACKSHIELD"]:
+        return shield
+    if tech_name in ["SHP_MULTISPEC_SHIELD"]:
+        return max(shield, stealth)
+
+    # ship armor TODO include these branches into ship-design calculation
+    if tech_name in ["SHP_ROOT_ARMOR", "SHP_ZORTRIUM_PLATE"]:
```

Comment:
Any X_ROOT_Y techs that are granted at the start of game (which I think is all of them, yes?) can be just left with the default score of 1 rather than cluttering up the decision making, can't they?